schism/compiler.ss

```diff
@@ -895,10 +1081,94 @@
      ((pair? exprs)
       (let ((expr (compile-expr (car exprs) env))
             (rest (compile-begin (cdr exprs) env)))
-        (if (eq? (caar exprs) '%store-mem)
+        (if (and (eq? (caar exprs) 'icall)
+                 (eq? (cadar exprs) '%store-mem))
             (cons expr rest)
             (cons `(drop ,expr) rest))))
      (else (error 'compile-begin "compile-begin: invalid begin"))))
+  (define (compile-icall op args env)
+    (cond
+     ((eq? op '%set-car!)
+      (let ((p (compile-expr (car args) env))
+            (x (compile-expr (cadr args) env)))
+        `(begin (i32.store (offset 0) (i32.and ,p (i32.const ,(fixnum-mask))) ,x)
+                (i32.const ,(constant-void)))))
+     ((eq? op '%set-cdr!)
+      (let ((p (compile-expr (car args) env))
+            (x (compile-expr (cadr args) env)))
+        `(begin (i32.store (offset ,(word-size)) (i32.and ,p (i32.const ,(fixnum-mask))) ,x)
+                (i32.const ,(constant-void)))))
+     ((eq? op 'eof-object) `(i32.const ,(constant-eof)))
+     ((eq? op '+)
+      (let ((a (compile-expr (car args) env))
+            (b (compile-expr (cadr args) env)))
+        `(i32.add ,a ,b)))
+     ((eq? op '-)
+      `(i32.sub ,(compile-expr (car args) env) ,(compile-expr (cadr args) env)))
+     ((eq? op '*)
+      (let ((a (compile-expr (car args) env))
+            (b (compile-expr (cadr args) env)))
+        ;; Shift only one of them and we don't have to shift back when we're done.
+        `(i32.mul (i32.shr_s ,a (i32.const ,(tag-size))) ,b)))
+     ((eq? op '/)
+      (let ((a (compile-expr (car args) env))
+            (b (compile-expr (cadr args) env)))
+        ;; Shift only one of them and we don't have to shift back when we're done.
```

Comment:
Huh, we shift both arguments, so I'm not sure the comment here is really valid.

Reply:
True!  I think this comment comes from the original code, fwiw.

Reply:
Yeah, it does. Probably sloppy copy and pasting on my part :)